sycl/test-e2e/Annotated_usm/annotated_usm_align_shared.cpp

```diff
@@ -0,0 +1,329 @@
+                           [&]() {
+                             return ATAnnotated(align, dev, Ctx, alloc::shared);
+                           }});
+
+  // Case: aligned_alloc_xxx<T> with compile-time alignment<K> (K is a power of
+  // 2), the alignment is the least-common-multiple of the alignment argument (a
+  // power of 2), K and sizeof(T)
+  CheckAlignLcmAll(
+      {{align, 32, sizeof(T)}, {align, 64, sizeof(T)}, {align, 512, sizeof(T)}},
+      std::tuple{[&]() { return ATShared(align, q, ALN32); },
+                 [&]() { return ATShared(align, dev, Ctx, ALN64); },
+                 [&]() {
+                   return ATAnnotated(align, dev, Ctx, alloc::shared, ALN512);
+                 }});
+
+  // Test cases that are expected to return null
+  auto check_null = [&q](auto AllocFn, int Line = __builtin_LINE(),
+                         int Case = 0) {
+    decltype(AllocFn()) Ptr = AllocFn();
+    auto v = reinterpret_cast<uintptr_t>(Ptr);
+    if (v != 0) {
+      free(Ptr, q);
+      std::cout << "Failed at line " << Line << ", case " << Case << std::endl;
+      assert(false && "The return is not null!");
+    }
+  };
+
+  auto CheckNullAll = [&](auto Funcs, int Line = __builtin_LINE()) {
+    std::apply(
+        [&](auto... Fs) {
+          int Case = 0;
+          (void)std::initializer_list<int>{
+              (check_null(Fs, Line, Case++), 0)...};
+        },
+        Funcs);
+  };
+
+  CheckNullAll(std::tuple{
+      // Case: malloc_xxx with compile-time alignment<K> (K is not a power of
+      // 2),
+      // nullptr is returned
+      [&]() { return MShared(q, properties{alignment<11>}); },
+      [&]() { return MShared(dev, Ctx, properties{alignment<13>}); },
+      [&]() { return MAnnotated(q, alloc::shared, properties{alignment<15>}); },
+      [&]() {
+        return MAnnotated(dev, Ctx, alloc::shared, properties{alignment<17>});
+      },
+      [&]() {
+        return MAnnotated(q, properties{usm_kind_shared, alignment<31>});
+      },
+      [&]() {
+        return MAnnotated(dev, Ctx, properties{usm_kind_shared, alignment<63>});
+      }
+
+      // Case: malloc_xxx<T> with compile-time alignment<K> (K is not a power of
+      // 2),
+      // nullptr is returned
+      ,
+      [&]() { return TShared(q, properties{alignment<11>}); },
+      [&]() { return TShared(dev, Ctx, properties{alignment<13>}); },
+      [&]() { return TAnnotated(q, alloc::shared, properties{alignment<15>}); },
+      [&]() {
+        return TAnnotated(dev, Ctx, alloc::shared, properties{alignment<17>});
+      },
+      [&]() {
+        return TAnnotated(q, properties{usm_kind_shared, alignment<31>});
+      },
+      [&]() {
+        return TAnnotated(dev, Ctx, properties{usm_kind_shared, alignment<63>});
+      }
+
+      // Case: aligned_alloc_xxx with no alignment property, and the alignment
+      // argument is not a power of 2, the result is nullptr
+      ,
+      [&]() { return AShared(11, q); },
```

Comment:
Can you add some tests with numbers that aren't odd, like 126?

Reply:
Remove the test cases for non-power-of-2 alignments based on the offline discussion

Reply:
Since SYCL is based on C++ and C++ only allows power-of-2 alignments, I think we should be able to test alignments that are not powers-of-2. The annotated malloc APIs should return annotated_ptr whose raw pointer is nullptr in this case.